File: src/main/java/org/distril/beengine/scheduler/Scheduler.java
package org.distril.beengine.scheduler;

import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import org.distril.beengine.scheduler.task.RunnableTask;
import org.distril.beengine.scheduler.task.Task;
import org.distril.beengine.scheduler.task.TaskEntry;

import java.util.ArrayDeque;
import java.util.Queue;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.TimeUnit;

public class Scheduler {

	private static final ForkJoinPool POOL = ForkJoinPool.commonPool();

	private final Queue<TaskEntry> queue = new ArrayDeque<>();

	private long lastUpdateTick;

	public TaskBuilder prepareTask(Runnable runnable) {
		return new TaskBuilder(this, new RunnableTask(runnable));
	}


	public TaskBuilder prepareTask(Task task) {
		return new TaskBuilder(this, task);
	}

	public void processTick(long currentTick) {
		this.lastUpdateTick = currentTick;
		while (!this.queue.isEmpty() && this.queue.peek().getNextRunTick() <= currentTick) {
			var taskEntry = this.queue.poll();
			if (!taskEntry.isCancelled()) {
				var task = taskEntry.getTask();
				if (taskEntry.isAsync()) {
					POOL.submit(task::onRun);
				} else {
					task.onRun();
				}

				if (!task.isCancelled() && taskEntry.isRepeating()) {
					this.addInQueue(taskEntry);
				}
			}
		}
	}

	protected void addInQueue(TaskEntry entry) {
		if (entry.isRepeating()) {
			entry.setNextRunTick(this.lastUpdateTick + entry.getPeriod());
		} else {
			entry.setNextRunTick(this.lastUpdateTick + entry.getDelay());
		}

		this.queue.add(entry);
	}

	public void cancelAllTasks() {
		this.queue.forEach(entry -> entry.getTask().cancel());
		this.queue.clear();

		POOL.shutdown();
		try {
			if (!POOL.awaitTermination(1, TimeUnit.SECONDS)) {
				POOL.shutdownNow();
			}
		} catch (InterruptedException exception) {
			POOL.shutdownNow();
		}
	}

	@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
	public static class TaskBuilder {

		private final Scheduler scheduler;
		private final Task task;

		private int delay, period;
		private boolean async;

		public TaskBuilder delay(int delay) {
			this.delay = delay;
			return this;
		}

		public TaskBuilder repeating(int period) {
			this.period = period;
			return this;
		}

		public TaskBuilder async(boolean async) {
			this.async = async;
			return this;
		}

		@SuppressWarnings("UnusedReturnValue")
		public Task schedule() {
			this.scheduler.addInQueue(new TaskEntry(this.task, this.delay, this.period, this.async));

			return this.task;
		}
	}
}

File: src/main/kotlin/org/distril/beengine/scheduler/Scheduler.kt
package org.distril.beengine.scheduler

import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout
import org.distril.beengine.scheduler.task.RunnableTask
import org.distril.beengine.scheduler.task.Task
import org.distril.beengine.scheduler.task.TaskEntry
import org.distril.beengine.util.Utils.getLogger
import java.util.concurrent.Executors
import java.util.concurrent.PriorityBlockingQueue
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicLong

class Scheduler(private val taskTimeout: Long) {

	private val executor = Executors.newCachedThreadPool()

	private val queue = PriorityBlockingQueue<TaskEntry>()

	private val lastUpdateTick = AtomicLong(0)

	fun processTick(currentTick: Long) {
		this.lastUpdateTick.set(currentTick)

		synchronized(this.queue) {
			while (this.queue.isNotEmpty() && this.queue.peek().nextRunTick <= currentTick) {
				val taskEntry = this.queue.poll()
				if (taskEntry.isAsync) {
					this.executor.submit { this.executeTask(taskEntry) }
				} else {
					this.executeTask(taskEntry)
				}
			}
		}
	}

	private fun executeTask(entry: TaskEntry) {
		if (!entry.isCancelled) {
			runBlocking {
				try {
					withTimeout(taskTimeout) { entry.task.onRun() }
				} catch (exception: TimeoutCancellationException) {
					log.error("Task timed out:", exception)
				} catch (exception: Exception) {
					log.error("Task failed with exception:", exception)
				}

				if (entry.isRepeating && !entry.isCancelled) addInQueue(entry, entry.period)
			}
		}
	}

	fun scheduleTask(delay: Int = 0, period: Int = 0, async: Boolean = false, runnable: Runnable) =
		this.scheduleTask(delay, period, async, RunnableTask(runnable))

	fun scheduleTask(delay: Int = 0, period: Int = 0, async: Boolean = false, task: Task) =
		this.addInQueue(TaskEntry(task, delay, period, async), delay)


	private fun addInQueue(entry: TaskEntry, adds: Int): TaskEntry {
		synchronized(this.queue) {
			entry.nextRunTick = this.lastUpdateTick.get() + adds
			this.queue.offer(entry)
		}

		return entry
	}

	fun cancelAllTasks() {
		synchronized(this.queue) {
			this.queue.forEach { it.cancel() }
			this.queue.clear()
		}

		this.executor.shutdown()
		try {
			if (!this.executor.awaitTermination(this.taskTimeout, TimeUnit.MILLISECONDS)) {
				this.executor.shutdownNow()
			}
		} catch (_: InterruptedException) {
			this.executor.shutdownNow()
		}
	}

	companion object {

		private val log = Scheduler.getLogger()
	}
}
